Sources/SwiftDocC/Semantics/DirectiveParser.swift

```diff
@@ -0,0 +1,66 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2024 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import Markdown
+
+/// A utlity type for parsing directives from markup.
+struct DirectiveParser<Directive: AutomaticDirectiveConvertible> {
+    
+    /// Returns a directive of the given type if found in the given sequence of markup elements and the remaining markup.
+    ///
+    /// If there are multiple instances of the same directive type, this functions returns the first instance
+    /// and diagnoses subsequent instances.
+    func parseSingleDirective<MarkupElements: Sequence<Markup>>(
+        _ directiveType: Directive.Type,
+        from markupElements: MarkupElements,
```

Comment:
That's a great suggestion, thanks! Initially I though we'd run into trouble because we sometimes call this function with `[any Markup]` and sometimes with `[any BlockDirective]` and I thought that'd be problematic with type erasure and `any Markup` not conforming to `Markup`, but we can actually just make this function take `inout [any Markup]`, so I did that.